anymail/webhooks/postal.py

```diff
@@ -0,0 +1,175 @@
+            )
+        except InvalidSignature:
+            raise AnymailWebhookValidationFailure(
+                "Postal webhook called with incorrect signature")
+
+
+class PostalTrackingWebhookView(PostalBaseWebhookView):
+    """Handler for Postal message, engagement, and generation event webhooks"""
+
+    signal = tracking
+
+    def parse_events(self, request):
+        esp_event = json.loads(request.body.decode("utf-8"))
+
+        raw_timestamp = esp_event.get("timestamp")
+        timestamp = (
+            datetime.fromtimestamp(int(raw_timestamp), tz=utc)
+            if raw_timestamp
+            else None
+        )
+
+        payload = esp_event.get("payload", {})
+
+        status_types = {
+            "Sent": EventType.DELIVERED,
+            "SoftFail": EventType.DEFERRED,
+            "HardFail": EventType.FAILED,
+            "Held": EventType.QUEUED,
+        }
+
+        def get_event_type(event):
+            if "status" in event:
+                return status_types.get(event["status"], EventType.UNKNOWN)
+
+            if "bounce" in event:
+                return EventType.BOUNCED
+
+            if "url" in event:
+                return EventType.CLICKED
+
+            return EventType.UNKNOWN
+
+        event_type = get_event_type(payload)
+
+        description = payload.get("details")
+        mta_response = payload.get("output")
+
+        # extract message-related fields
+        message = payload.get("message") or payload.get("original_message", {})
+        message_id = message.get("message_id")
+        tag = message.get("tag")
+        recipient = None
+        message_to = message.get("to")
+        if message_to is not None:
+            try:
+                recipient = parse_single_address(message_to).addr_spec
+            except AnymailInvalidAddress:
+                pass
+
+        if message.get("direction") == "incoming":
+            # This is a tracking event about inbound email, so it's not interesting to anymail.
+            # We can safely ignore this.
+            return []
+
+        # only for MessageLinkClicked
+        click_url = payload.get("url")
+        user_agent = payload.get("user_agent")
+
+        event = AnymailTrackingEvent(
+            event_type=event_type,
+            timestamp=timestamp,
+            event_id=None,
+            esp_event=esp_event,
+            click_url=click_url,
+            description=description,
+            message_id=message_id,
+            metadata=None,
+            mta_response=mta_response,
+            recipient=recipient,
+            reject_reason=None,  # Postal doesn't provide a reject reason
+            tags=[tag],
+            user_agent=user_agent,
+        )
+
+        return [event]
+
+
+class PostalInboundWebhookView(PostalBaseWebhookView):
+    """Handler for Postal inbound relay webhook"""
+
+    signal = inbound
+
+    def parse_events(self, request):
+        esp_event = json.loads(request.body.decode("utf-8"))
+
```

Comment:
Similar to earlier comment, if there's a field you can check to make sure this actually *is* an inbound event, and issue a ConfigurationError if not, that's helpful. (If not, don't worry about it.)

Reply:
Done!